File: prototype/prototype.cs
namespace ConsoleApp1;

abstract class Prototype
{
    public abstract Prototype Clone();

    public void Show()
    {
        Console.WriteLine(this.GetType().Name);
    }
}

class ConcretePrototypeA : Prototype
{
    public override ConcretePrototypeA Clone()
    {
        return new ConcretePrototypeA();
    }
}

class ConcretePrototypeB : Prototype
{
    public override ConcretePrototypeB Clone()
    {
        return new ConcretePrototypeB();
    }
}

class Client
{
    private Prototype _prototype;

    public Client(Prototype prototype)
    {
        this._prototype = prototype;
    }

    public Prototype GetProduct()
    {
        return this._prototype.Clone();
    }
}

class TestClass {}

class Program
{
    static void Main()
    {
        ConcretePrototypeA prototype = new();
        Client client = new(prototype);
        Prototype product = client.GetProduct();
        product.Show();
    }
}

File: prototype/prototype.py
from abc import ABC, abstractmethod
from typing import Self


class Prototype(ABC):
    @abstractmethod
    def clone(self) -> Self:
        raise NotImplementedError

    def __str__(self) -> str:
        return f"{type(self).__name__}"


class ConcretePrototypeA(Prototype):
    def clone(self) -> Self:
        return type(self)()


class ConcretePrototypeB(Prototype):
    def clone(self) -> Self:
        return type(self)()


class Client:
    def __init__(self, prototype: Prototype) -> None:
        self.prototype = prototype

    def get_product(self) -> Prototype:
        return self.prototype.clone()


def main() -> None:
    prototype = ConcretePrototypeB()
    client = Client(prototype)
    product = client.get_product()
    print(product)


if __name__ == "__main__":
    main()
